service/controller/app/v1/resource/kubeconfigfinalizer/delete.go

```diff
@@ -0,0 +1,65 @@
+package kubeconfigfinalizer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/app-operator/service/controller/app/v1/key"
+)
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if key.InCluster(cr) {
+		return nil
+	}
+
+	name := key.KubecConfigSecretName(cr)
+	namespace := key.KubecConfigSecretNamespace(cr)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding kubeconfig secret for app %#q", name))
+
+	kubeConfig, err := r.k8sClient.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find kubeconfig %#q", name))
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found kubeconfig secret for app %#q", name))
+
+	finalizerTag := key.KubeConfigFinalizer(cr)
+
+	if contains(kubeConfig.Finalizers, finalizerTag) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("clear finalizer for kubeconfig %#q", name))
+
+		kubeConfig.Finalizers = filter(kubeConfig.Finalizers, finalizerTag)
+
+		_, err := r.k8sClient.CoreV1().Secrets(namespace).Update(kubeConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("kubeconfig %#q do not have matching finalizer", name))
+	}
```

Comment:
I wonder if this is ok to just remove this finalizer here. Are you sure access to the TC isn't needed when we reach this resource? What if the Chart CR deletion on the TC is still in progress? I think there should be some sort of communication between resources or reconciliation cancellation.

Reply:
Is it possible other resources are in progress still when we tried to delete finalizer in here? 
I thought we put this `kubeconfigfinalizer` at the last step so there should be no remaining tasks.

Reply:
No but other resources should not block. They may cancel operation if Chart CR deletion is in progress.

Reply:
the operator does not check chart CR deletion in progress cases yet. I can add in `chart/current.go` to check CR deletion status and could cancel reconciliation at that point. 
But I would like to keep this PR as smaller as possible, Do you mind if I would start this after this PR merged?

Reply:
We should start to think about a concept of reconciliation result so that we have a standardized way of communicating state of resource implementations at runtime. Our current primitive is the controller context but that is ugly and a bit unstructured.

Reply:
Okay but let's keep this in mind.